File: apps/api/src/lib/tokens.ts
import { createHash, randomBytes } from "node:crypto";
import type { UserRole } from "@prisma/client";
import { type JWTPayload, SignJWT, jwtVerify } from "jose";

export const ACCESS_COOKIE = "access_token";
export const REFRESH_COOKIE = "refresh_token";

export type AccessTokenPayload = {
  sub: string;
  email: string;
  role: UserRole;
};

export function hashToken(token: string): string {
  return createHash("sha256").update(token).digest("hex");
}

export function createRefreshTokenValue(): string {
  return randomBytes(48).toString("base64url");
}

export async function signAccessToken(
  payload: AccessTokenPayload,
  secret: string,
  expiresIn = "15m",
): Promise<string> {
  return new SignJWT({
    email: payload.email,
    role: payload.role,
  })
    .setProtectedHeader({ alg: "HS256" })
    .setSubject(payload.sub)
    .setIssuedAt()
    .setExpirationTime(expiresIn)
    .sign(new TextEncoder().encode(secret));
}

function readStringClaim(payload: JWTPayload, claim: string): string | undefined {
  const value = payload[claim];
  return typeof value === "string" ? value : undefined;
}

export async function verifyAccessToken(
  token: string,
  secret: string,
): Promise<AccessTokenPayload> {
  const { payload } = await jwtVerify(token, new TextEncoder().encode(secret));
  const sub = payload.sub;
  const email = readStringClaim(payload, "email");
  const role = readStringClaim(payload, "role");
  if (typeof sub !== "string" || !email || (role !== "user" && role !== "admin")) {
    throw new Error("Invalid access token payload");
  }
  return { sub, email, role };
}

File: templates/api-python/app/lib/tokens.py
from __future__ import annotations

import hashlib
import secrets
from datetime import UTC, datetime, timedelta
from typing import Any, Literal

import jwt

ACCESS_COOKIE = "access_token"
REFRESH_COOKIE = "refresh_token"

UserRole = Literal["user", "admin"]


def hash_token(token: str) -> str:
    return hashlib.sha256(token.encode("utf-8")).hexdigest()


def create_refresh_token_value() -> str:
    return secrets.token_urlsafe(48)


def sign_access_token(
    *,
    sub: str,
    email: str,
    role: UserRole,
    secret: str,
    expires_minutes: int = 15,
    now: datetime | None = None,
) -> str:
    issued_at = now or datetime.now(UTC)
    if issued_at.tzinfo is None:
        issued_at = issued_at.replace(tzinfo=UTC)
    payload: dict[str, Any] = {
        "sub": sub,
        "email": email,
        "role": role,
        "iat": issued_at,
        "exp": issued_at + timedelta(minutes=expires_minutes),
    }
    encoded = jwt.encode(payload, secret, algorithm="HS256")
    return encoded if isinstance(encoded, str) else encoded.decode("utf-8")


def verify_access_token(token: str, secret: str) -> dict[str, str]:
    payload = jwt.decode(token, secret, algorithms=["HS256"])
    sub = payload.get("sub")
    email = payload.get("email")
    role = payload.get("role")
    if not isinstance(sub, str) or not isinstance(email, str) or role not in ("user", "admin"):
        raise ValueError("Invalid access token payload")
    return {"sub": sub, "email": email, "role": role}
